xds/src/main/java/io/grpc/xds/internal/certprovider/ZatarCertificateProvider.java

```diff
@@ -0,0 +1,189 @@
+    }
+    getWatcher().close();
+  }
+
+  private void scheduleNextRefreshCertificate(long delayInSeconds) {
+    if (scheduledHandle != null && scheduledHandle.isPending()) {
+      logger.log(Level.SEVERE, "Pending task found: inconsistent state in scheduledHandle!");
+      scheduledHandle.cancel();
+    }
+    RefreshCertificateTask runnable = new RefreshCertificateTask();
+    scheduledHandle =
+        syncContext.schedule(runnable, delayInSeconds, TimeUnit.SECONDS, scheduledExecutorService);
+  }
+
+  @VisibleForTesting
+  void checkAndReloadCertificates() {
+    try {
+      Path targetPath = Files.readSymbolicLink(directory);
+      if (targetPath.equals(lastModifiedTarget)) {
+        return;
+      }
+      PrivateKey privateKey =
+          CertificateUtils.getPrivateKey(
+              new FileInputStream(new File(targetPath.toFile(), privateKeyFile)));
+      X509Certificate[] certs =
+          CertificateUtils.toX509Certificates(
+              new FileInputStream(new File(targetPath.toFile(), certFile)));
+      X509Certificate[] caCerts =
+          CertificateUtils.toX509Certificates(
+              new FileInputStream(new File(targetPath.toFile(), trustFile)));
+      getWatcher().updateCertificate(privateKey, Arrays.asList(certs));
+      getWatcher().updateTrustedRoots(Arrays.asList(caCerts));
+      lastModifiedTarget = targetPath;
+    } catch (Throwable t) {
+      generateErrorIfCurrentCertExpired(t);
+    } finally {
+      scheduleNextRefreshCertificate(refreshIntervalInSeconds);
+    }
+  }
+
+  private void generateErrorIfCurrentCertExpired(Throwable t) {
+    X509Certificate currentCert = getWatcher().getLastIdentityCert();
+    if (currentCert != null) {
+      long delaySeconds = computeDelaySecondsToCertExpiry(currentCert);
+      if (delaySeconds > GRACE_INTERVAL_IN_SECONDS) {
+        logger.log(Level.FINER, "reload certificate error", t);
+        return;
+      }
+      getWatcher().clearValues();
+    }
+    getWatcher().onError(Status.fromThrowable(t));
+  }
+
+  @SuppressWarnings("JdkObsolete")
+  private long computeDelaySecondsToCertExpiry(X509Certificate lastCert) {
+    checkNotNull(lastCert, "lastCert");
+    return TimeUnit.NANOSECONDS.toSeconds(
+        TimeUnit.MILLISECONDS.toNanos(lastCert.getNotAfter().getTime())
+            - timeProvider.currentTimeNanos());
+  }
+
+  @VisibleForTesting
+  class RefreshCertificateTask implements Runnable {
+    @Override
+    public void run() {
+      checkAndReloadCertificates();
+    }
+  }
+
+  private final SynchronizationContext syncContext;
+  private final ScheduledExecutorService scheduledExecutorService;
+  private final TimeProvider timeProvider;
+  private final Path directory;
+  private final String certFile;
+  private final String privateKeyFile;
+  private final String trustFile;
+  private final long refreshIntervalInSeconds;
+  @VisibleForTesting SynchronizationContext.ScheduledHandle scheduledHandle;
+  private Path lastModifiedTarget;
```

Comment:
I'd prefer to move the fields up to the topmost to be somewhat consistent in style.